README.rst

```diff
@@ -471,6 +471,120 @@ any application code:
+             params={'*stream':'"stdout"'} , output = 'ruleExecOut' )
+  output = r.execute( )
+  if output and len(output.MsParam_PI):
+      buf = output.MsParam_PI[0].inOutStruct.stdoutBuf.buf
+      if buf: print(buf.rstrip(b'\0').decode('utf8'))
+
+(Changing the input value to be squared in this example is left as an exercise for the reader!)
+
+To deal with errors resulting from rule execution failure, two approaches can be taken. Suppose we
+have defined this in the :code:`/etc/irods/core.re` rule-base::
+
+  rule_that_fails_with_error_code(*x) {
+    *y = (if (*x!="") then int(*x) else 0)
+  # if (SOME_PROCEDURE_GOES_WRONG) {
+      if (*y < 0) { failmsg(*y,"-- my error message --"); }  #-> throws an error code of int(*x) in REPF
+      else { fail(); }                                       #-> throws FAIL_ACTION_ENCOUNTERED_ERR in REPF
+  # }
+  }
+
+We can run the rule thus:
+
+>>> Rule( session, body='rule_that_fails_with_error_code(""), instance_name = 'irods_rule_engine_plugin-irods_rule_language-instance',
+        ).execute( r_error = (r_errs:= irods.message.RErrorStack()) )
+
+Where we've used the Python 3.8 "walrus operator" for brevity.  The error will automatically be caught and translated to a
+returned-error stack::
+
+   >>> pprint.pprint([vars(r) for r in r_errs])
+   [{'raw_msg_': 'DEBUG: fail action encountered\n'
+                 'line 14, col 15, rule base core\n'
+                 '        else { fail(); }\n'
+                 '               ^\n'
+                 '\n',
+     'status_': -1220000}]
+
+Note, if a stringized negative integer is given , ie. as a special fail code to be thrown within the rule,
+we must add this code into a special parameter to have this automatically caught as well:
+
+>>> Rule( session, body='rule_that_fails_with_error_code("-2")',instance_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+...     ).execute( r_error = (r_errs := irods.message.RErrorStack()),
+...           I    acceptable_errors = ( FAIL_ACTION_ENCOUNTERED_ERR, -2))
+
+Because rule is written to emit a custom error message via failmsg int this case, the resulting r_error stack will now include that
+custom error message as a substring::
+
+  >>> pprint.pprint([vars(r) for r in r_errs])
+  [{'raw_msg_': 'DEBUG: -- my error message --\n'
+              'line 21, col 20, rule base core\n'
+              '      if (*y < 0) { failmsg(*y,"-- my error message --"); }  '
+              '#-> throws an error code of int(*x) in REPF\n'
+              '                    ^\n'
+              '\n',
+    'status_': -1220000}]
+
+Alternatively or in combination with the automatic catching of errors, we may also catch errors as exceptions on the client
+side.  If e.g. the Python rule engine is configured with the following rule in :code:`/etc/irods/core.py`::
+
+  def python_rule(rule_args, callback, rei):
+  #   if some operation fails():
+          raise RuntimeError
```

Comment:
whitespace - yeah, just make sure all these example code blocks render as expected.

Reply:
This was a way of indicating that when something goes wrong, we can raise a RuntimeError  to fail in a python rule. This was the indentation I intended, because the comment could be a real statement if needed.  But it also works as-is.     I don't know, maybe it looks off-putting to some eyes....

Reply:
no, i mean, the left-edge indentation.   it's not the same as other code blocks.

the indentation within the code itself is fine and up to you.